pkg/controller/networkpolicy/mutate.go

```diff
@@ -0,0 +1,201 @@
+	networkPolicyController *NetworkPolicyController
+}
+
+// NewNetworkPolicyMutator returns a new *NetworkPolicyMutator.
+func NewNetworkPolicyMutator(networkPolicyController *NetworkPolicyController) *NetworkPolicyMutator {
+	return &NetworkPolicyMutator{
+		networkPolicyController: networkPolicyController,
+	}
+}
+
+// Mutate function mutate a Antrea Policy object
+func (m *NetworkPolicyMutator) Mutate(ar *admv1.AdmissionReview) *admv1.AdmissionResponse {
+	var result *metav1.Status
+	var msg string
+	var patch []byte
+	allowed := false
+	patchType := admv1.PatchTypeJSONPatch
+
+	op := ar.Request.Operation
+	curRaw := ar.Request.Object.Raw
+	oldRaw := ar.Request.OldObject.Raw
+
+	switch ar.Request.Kind.Kind {
+	case "ClusterNetworkPolicy":
+		klog.V(2).Info("Mutating Antrea ClusterNetworkPolicy CRD")
+		var curACNP, oldACNP secv1alpha1.ClusterNetworkPolicy
+		if curRaw != nil {
+			if err := json.Unmarshal(curRaw, &curACNP); err != nil {
+				klog.Errorf("Error de-serializing current Antrea ClusterNetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		if oldRaw != nil {
+			if err := json.Unmarshal(oldRaw, &oldACNP); err != nil {
+				klog.Errorf("Error de-serializing old Antrea ClusterNetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		msg, allowed, patch = m.mutateAntreaPolicyRuleName(op, curACNP.Spec.Ingress, curACNP.Spec.Egress)
+	case "NetworkPolicy":
+		klog.V(2).Info("Mutating Antrea NetworkPolicy CRD")
+		var curANP, oldANP secv1alpha1.NetworkPolicy
+		if curRaw != nil {
+			if err := json.Unmarshal(curRaw, &curANP); err != nil {
+				klog.Errorf("Error de-serializing current Antrea NetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		if oldRaw != nil {
+			if err := json.Unmarshal(oldRaw, &oldANP); err != nil {
+				klog.Errorf("Error de-serializing old Antrea NetworkPolicy")
+				return GetAdmissionResponseForErr(err)
+			}
+		}
+		msg, allowed, patch = m.mutateAntreaPolicyRuleName(op, curANP.Spec.Ingress, curANP.Spec.Egress)
+	}
+
+	if msg != "" {
+		result = &metav1.Status{
+			Message: msg,
+		}
+	}
+	return &admv1.AdmissionResponse{
+		Allowed:   allowed,
+		Result:    result,
+		PatchType: &patchType,
+		Patch:     patch,
+	}
+}
+
+// mutateAntreaPolicyRuleName mutates names of rules of an Antrea NetworkPolicy CRD.
+// If users didn't specify the name of an ingress or egress rule,
+// mutateAntreaPolicyRuleName will auto-generate a name for this rule.
+func (m *NetworkPolicyMutator) mutateAntreaPolicyRuleName(op admv1.Operation, ingress, egress []secv1alpha1.Rule) (string, bool, []byte) {
+	allowed := true
+	reason := ""
+	var patch []byte
+	switch op {
+	case admv1.Create, admv1.Update:
+		ruleNames := sets.NewString()
+		for _, rule := range append(ingress, egress...) {
+			ruleNames.Insert(rule.Name)
+		}
+
+		ingressPath, ingressValue := generateName(ruleNames, "ingress", ingress)
+		egressPath, egressValue := generateName(ruleNames, "egress", egress)
```

Comment:
names should be `ingressRulePaths` and `ingressRuleNames` I think

same for egress

Reply:
Thx.